pandas/core/internals/blocks.py

```diff
@@ -2066,7 +2067,7 @@ def _can_hold_element(self, element: Any) -> bool:
                 and not issubclass(tipo.type, (np.datetime64, np.timedelta64))
                 and self.dtype.itemsize >= tipo.itemsize
             )
-        return is_integer(element)
+        return is_integer(element) or (is_float(element) and element.is_integer())
```

Comment:
this would only hit a non-inferred path, i don't think your added code does anything.

Reply:
If the element is a simple python float, `maybe_infer_dtype_type` return `None` so we are in this case.